code_docs/DirectoryReference.md

```diff
@@ -0,0 +1,94 @@
+
+# Source Folder 
+`DelugeFirmware/src/`
+
+*Contains all source code. Most functional code lives in subfolder [`deluge/`](#deluge-folder). Everything else is managing low-level systems, global definitions, etc.*
+
+# Deluge Folder
+`DelugeFirmware/src/deluge`
+
+*Main directory for firmware code. `deluge.cpp` runs the boot sequence and main loop*
+
+| Folder | Contents
+|-|-
+| drivers | mostly deals with managing hardware
+| dsp | digital signal processing, manages audio effects
+| [gui](#gui-folder) | graphical user interface, manages all user-facing behavior of the Deluge
+| hid | human interface device, manages basic functionality of knobs, buttons, pads, LEDs, display
+| io | input/output, manages debug-logging and incoming midi
+| memory | memory management
+| [model](#model-folder) | manages model_stack, the model Deluge uses to represent and pass around its internal state
+| modulation | manages internal modulating of parameters, including automation and midi
+| playback | manages playback and recording modes
+| processing | manages various processes: audio, CV and live audio engines, sound instruments, metronome, stem export
+| storage | manages using files from the SD card
+| testing | automated tests
+| util | collection of general utilities
+| version | deals with the code version
+
+## GUI Folder
+`DelugeFirmware/src/deluge/gui`
+
+*Graphical User Interface. This folder manages all user-facing behavior of the Deluge and is therefor the main focus for fixing bugs and developing new features*
+
+| Folder | Contents
+|-|-
+| colour | colour definitions
+| context_menu | the context menu pops up on top of the regular menu, depending on contextual actions
+| fonts | font definitions
+| l10n | list of strings used in the ui for both 7SEG and oled, can be used for localization
+| menu_item | definitions for the options of all menus
+| ui | manages the foundational layer of the ui including menus, browsers, keyboard views and saving/loading. `ui.cpp` serves as a base-class that the different views/modes of the Deluge are built on
+| [views](#views-folder) | manages the main views/modes of the Deluge, such as clip view and arranger view
+| waveform | manages the waveform view
+
+### UI Folder
+`DelugeFirmware/src/deluge/gui/ui/`
+
+### Views Folder
+`DelugeFirmware/src/deluge/gui/views`
+
+*This folder manages all the main modes of the Deluge ui*
+
+Each of the `final` views - that is, each view that is not meant to be a base-class for something else - defines a global variable of itself at the bottom of its header file. For example:
+```cpp
+extern AutomationView automationView;
+```
+This global variable is reused anytime that view is activated.
+
+| View | Info
+|-|-
+| `ArrangerView final` | is a `TimelineView` representing the arranger view
+| `AudioClipView final` | is a `ClipView` for audio clips
+| `AutomationView final` | is a `ClipView` that handles the automation layer of a clip
```

Comment:
Funny enough this is revealing of a change that needs happen in the refactoring I eventually want to do. AutomationView is no longer just a clip automation view but it's also an arranger automation view. So to a reader / new contributor I can see how this is confusing.

Reply:
This needs a solution in one of a few ways, to reuse behavior in different scenarios:
- View as base class, ScrollingView as something like a template that can be implemented on views that work with the arranger
- View as base class with a toggle to allow arranger-like scrolling where necessary
- We could go absolutely wild and implement free scrolling on all views, while we're at it. (I honestly don't see any reason against it? Should of course be toggle-able to classic behavior.)